### PR TITLE
Integrate AutoModelForSequenceClassification through PytorchModel

### PR DESCRIPTION
**Pull request by abarbosa94**
### Description
- This PR adds an initial implementation for the HuggingFace SequenceClassification model with some initial functionality. I also organized imports using [isort](https://pypi.org/project/isort/) and made a few adjustments to increase flake8 compliance

This should be taken as an initial step toward: #238, #103, and #217 

### Implemented changes
- Insert a description of the changes implemented in the pull request.
  - [x] Modify the `PyTorchModel` prediction method to accept the HuggingFace model
  - [x] Adds new tests that prove new feature
  - [x] Add flake8 parametrization to tox 
  - [x] Modify some import ordering according to `isort`

### Minimum acceptance criteria
- Specify what is necessary for the PR to be merged with the main branch.
- @mentions of the person that is apt to review these changes, e.g., @annahedstroem

### CODE REVIEW

File: quantus/helpers/model/pytorch_model.py [hunk 1]
```diff
@@ -97,6 +98,19 @@ def _get_model_with_linear_top(self) -> torch.nn:
...
         return linear_model
 
+    def _obtain_predictions(self, x, model_predict_kwargs):
+        pred = None
+        if isinstance(self.model, PreTrainedModel):
+            if not isinstance(x, BatchEncoding):
+                raise ValueError(
+                    "When using HuggingFace pretrained models, please use Tokenizers output for `x`"
+                )
+            pred = self.model(**x, **model_predict_kwargs).logits
```
Comment (annahedstroem) [inline quantus/helpers/model/pytorch_model.py:108]: Should we also enable softmax here (post accessing the logits)? so that we convert the pred to softmax, if softmax=True? (we can add it as a class attribute above).

File: pyproject.toml [hunk 2]
```diff
@@ -36,7 +36,9 @@ dependencies = [
     "scipy>=1.7.3",
     "tqdm>=4.62.3",
     "matplotlib>=3.3.4",
-    "typing_extensions; python_version <= '3.8'"
+    "typing_extensions; python_version <= '3.8'",
+    "transformers<=4.30.2; python_version == '3.7'",
```
Comment (annahedstroem) [inline pyproject.toml:40]: I wonder if this should be a part of the base dependencies. I think it is better fitted to add it under 'torch' (see line 78/80 and below)?

File: quantus/helpers/model/pytorch_model.py (hunk 1 above)
Comment (abarbosa94) [reply quantus/helpers/model/pytorch_model.py:119]: I implemented this way, please see if you agree: [9a67c4c](https://github.com/understandable-machine-intelligence-lab/Quantus/pull/339/commits/9a67c4c933784c3db962e9969bf74f4b409a12de)

File: quantus/helpers/model/pytorch_model.py [hunk 3]
```diff
@@ -97,6 +98,19 @@ def _get_model_with_linear_top(self) -> torch.nn:
 
         return linear_model
 
+    def _obtain_predictions(self, x, model_predict_kwargs):
+        pred = None
+        if isinstance(self.model, PreTrainedModel):
+            if not isinstance(x, BatchEncoding):
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply quantus/helpers/model/pytorch_model.py]: Done: [885a3e2](https://github.com/understandable-machine-intelligence-lab/Quantus/pull/339/commits/885a3e2fcd9cd32c3bbde3c6b80408d27a93795a)

File: pyproject.toml [hunk 4]
```diff
@@ -52,6 +52,7 @@ dynamic = ["version"]
 #
 [project.optional-dependencies]
 tests = [
+    "cachetools>=5.3.3",
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply pyproject.toml:56]: You're right, removing it

File: tox.ini [hunk 5]
```diff
@@ -66,3 +66,7 @@ python =
     3.9 = py39
     3.10 = py310
     3.11 = py311
+
+[flake8]
+max-line-length = 127
+max-complexity = 10
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tox.ini:72]: tbh I just noticed it is currently set [here](https://github.com/understandable-machine-intelligence-lab/Quantus/blob/2f19b31e01f6788ff5ae999232e110821f3af261/tox.ini#L45) and I wanted to make it more explicit

File: tox.ini [hunk 6]
```diff
@@ -66,3 +66,7 @@ python =
     3.9 = py39
     3.10 = py310
     3.11 = py311
+
+[flake8]
+max-line-length = 127
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tox.ini:71]: Considering it is set [here](https://github.com/understandable-machine-intelligence-lab/Quantus/blob/2f19b31e01f6788ff5ae999232e110821f3af261/tox.ini#L45), a lot of files are using the 127 line length (I tried to run black and noticed it). Can we change this as a technical debt in a specific PR? Otherwise this PR will be big and with a different context.

File: quantus/helpers/model/pytorch_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abarbosa94) [reply quantus/helpers/model/pytorch_model.py:119]: If we do this, dont it increase the number of returns redundantly? say currently we have:
```
          pred = self.model(**x, **model_predict_kwargs).logits
          if self.softmax:
              pred = torch.softmax(pred, dim=-1)
      elif isinstance(self.model, nn.Module):
          pred_model = self.get_softmax_arg_model()
          pred = pred_model(torch.Tensor(x).to(self.device), **model_predict_kwargs)
      return pred
```

it would then become:
```
          if self.softmax:
              return torch.softmax(self.model(**x, **model_predict_kwargs).logits
, dim=-1)
      return self.model(**x, **model_predict_kwargs).logits
      elif isinstance(self.model, nn.Module):
          pred_model = self.get_softmax_arg_model()
         return pred_model(torch.Tensor(x).to(self.device), **model_predict_kwargs)
```

what do you think?

File: quantus/helpers/model/pytorch_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abarbosa94) [reply quantus/helpers/model/pytorch_model.py]: I changed this behaviour, now it checks if the keys exists in the structure :)

File: quantus/helpers/model/pytorch_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abarbosa94) [reply quantus/helpers/model/pytorch_model.py:119]: I did slightly different to handle the softmax param and raise it properly. Could you see if you agree? Thanks

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (abarbosa94) [reply pyproject.toml]: This si a great question :)

File: tests/functions/test_pytorch_model.py [hunk 7]
```diff
@@ -242,3 +241,55 @@ def test_add_mean_shift_to_first_layer(load_mnist_model):
...
+        (
+            lazy_fixture("load_hf_distilbert_sequence_classifier"),
+            np.array([1, 2, 3]),
+            False,
+            {},
+            ValueError,
+        ),
+    ],
+)
+def test_huggingface_classifier_predict(hf_model, data, softmax, model_kwargs, expected):
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tests/functions/test_pytorch_model.py:289]: Done: 179da1e

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (abarbosa94) [reply pyproject.toml]: Done: 179da1e

File: quantus/helpers/model/pytorch_model.py [hunk 8]
```diff
@@ -97,6 +98,19 @@ def _get_model_with_linear_top(self) -> torch.nn:
...
+                raise ValueError(
+                    "When using HuggingFace pretrained models, please use Tokenizers output for `x`"
+                )
+            pred = self.model(**x, **model_predict_kwargs).logits
+        elif isinstance(self.model, nn.Module):
+            pred_model = self.get_softmax_arg_model()
+            pred = pred_model(torch.Tensor(x).to(self.device), **model_predict_kwargs)
+        return pred
+
     def get_softmax_arg_model(self) -> torch.nn:
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply quantus/helpers/model/pytorch_model.py]: Done: 179da1e

File: tests/functions/test_pytorch_model.py [hunk 9]
```diff
@@ -242,3 +241,37 @@ def test_add_mean_shift_to_first_layer(load_mnist_model):
...
+        ),
+        (
+            lazy_fixture("load_hf_distilbert_sequence_classifier"),
+            np.array([1, 2, 3]),
+            {},
+            ValueError,
+        ),
+    ],
+)
+def test_huggingface_classifier_predict(hf_model, data, model_kwargs, expected):
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tests/functions/test_pytorch_model.py]: Done: test_huggingface_classifier_predict

File: tests/conftest.py [hunk 10]
```diff
@@ -1,26 +1,26 @@
...
+from transformers import (AutoModelForSequenceClassification, AutoTokenizer,
+                          set_seed)
 
 CIFAR_IMAGE_SIZE = 32
 MNIST_IMAGE_SIZE = 28
 BATCH_SIZE = 124
 MINI_BATCH_SIZE = 8
+RANDOM_SEED = 42
+
+set_seed(42)
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tests/conftest.py:23]: `set_seed` from huggingface ensure all of these (and some others as well), but using `autouse` is a clever idea :) just did it!

File: tests/conftest.py [hunk 11]
```diff
@@ -236,6 +236,29 @@ def load_mnist_model_softmax():
...
+    """
+    DISTILBERT_BASE = "distilbert-base-uncased"
+    model = AutoModelForSequenceClassification.from_pretrained(
+        DISTILBERT_BASE, cache_dir="/tmp/"
+    )
+    return model
+
+
+@pytest.fixture(scope="session", autouse=True)
+def mock_hf_text():
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tests/conftest.py:252]: Haha, good catch! Just renamed it

File: tests/conftest.py [hunk 12]
```diff
@@ -236,6 +236,29 @@ def load_mnist_model_softmax():
...
+    TODO
+    """
+    DISTILBERT_BASE = "distilbert-base-uncased"
+    model = AutoModelForSequenceClassification.from_pretrained(
+        DISTILBERT_BASE, cache_dir="/tmp/"
+    )
+    return model
+
+
+@pytest.fixture(scope="session", autouse=True)
```
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tests/conftest.py]: Sorry, just did it

File: tests/conftest.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tests/conftest.py]: [9c56c50](https://github.com/understandable-machine-intelligence-lab/Quantus/pull/339/commits/9c56c50fce26c79b9de139a4dad5d98c72c2589b)

File: tests/conftest.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (abarbosa94) [reply tests/conftest.py]: [9c56c50](https://github.com/understandable-machine-intelligence-lab/Quantus/pull/339/commits/9c56c50fce26c79b9de139a4dad5d98c72c2589b)

File: tox.ini (hunk 6 above)
[the remark this answers is not in the document]
Comment (annahedstroem) [reply tox.ini:71]: agree -- this is definitely a topic for separate PR :)

File: quantus/helpers/model/pytorch_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/helpers/model/pytorch_model.py:120]: I think that looks great :D @abarbosa94

**Comment by annahedstroem** [comment]
based on the tesing, it looks like we need to add transformers also to the `tests` and not only `full`: https://github.com/understandable-machine-intelligence-lab/Quantus/actions/runs/8328433263/job/22788366589?pr=339

File: pyproject.toml [hunk 13]
```diff
@@ -85,7 +85,9 @@ torch = [
     "torchvision<=0.12.0; python_version == '3.7'",
     "torchvision>=0.15.1; sys_platform != 'linux' and python_version > '3.7'",
     "torchvision>=0.14.0, <0.15.1; sys_platform == 'linux' and python_version > '3.7' and python_version <= '3.10'",
-    "torchvision>=0.15.1; sys_platform == 'linux' and python_version >= '3.11'"
+    "torchvision>=0.15.1; sys_platform == 'linux' and python_version >= '3.11'",
+    "transformers<=4.30.2; python_version == '3.7'",
```
Comment (annahedstroem) [inline pyproject.toml:89]: can we add this to `tests`?

Comment (abarbosa94) [reply pyproject.toml:64]: Just moved tests to the end

File: quantus/helpers/model/pytorch_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aaarrti) [reply quantus/helpers/model/pytorch_model.py:120]: you're right, it looks a bit different now. Can we also remove `pred = None` at the top?

File: tests/functions/test_pytorch_model.py [hunk 14]
```diff
@@ -242,3 +242,52 @@ def test_add_mean_shift_to_first_layer(load_mnist_model):
...
+            lazy_fixture("load_hf_distilbert_sequence_classifier"),
+            np.array([1, 2, 3]),
+            False,
+            {},
+            pytest.raises(ValueError),
+        ),
+    ],
+)
+def test_huggingface_classifier_predict(hf_model, data, softmax, model_kwargs, expected):
+    model = PyTorchModel(model=hf_model, softmax=softmax, model_predict_kwargs=model_kwargs)
```
Comment (aaarrti) [inline tests/functions/test_pytorch_model.py:290]: I though `softmax` must be a bool, or?

File: tests/conftest.py [hunk 15]
```diff
@@ -236,7 +238,30 @@ def load_mnist_model_softmax():
...
+    TODO
+    """
+    DISTILBERT_BASE = "distilbert-base-uncased"
+    model = AutoModelForSequenceClassification.from_pretrained(
+        DISTILBERT_BASE, cache_dir="/tmp/"
+    )
+    return model
+
+
+@pytest.fixture(scope="session", autouse=False)
```
Comment (aaarrti) [inline tests/conftest.py:253]: `autouse=False` is the default

File: quantus/helpers/model/pytorch_model.py [hunk 16]
```diff
@@ -6,16 +6,19 @@
...
-import logging
+from functools import lru_cache
+from typing import Any, Dict, Generator, List, Mapping, Optional, Tuple, Union
 
 import numpy as np
+import numpy.typing as npt
 import torch
 from torch import nn
-from functools import lru_cache
+from transformers import PreTrainedModel
```
Comment (aaarrti) [inline quantus/helpers/model/pytorch_model.py:20]: This will cause `ModuleNotFoundError` when user tries to import `Quantus` without `transformers` installed.

File: pyproject.toml [hunk 17]
```diff
@@ -104,8 +81,39 @@ zennit = [
     "quantus[torch]",
     "zennit>=0.5.1"
 ]
+transformers = [
+    "quantus[torch, tensorflow]",
```
Comment (aaarrti) [inline pyproject.toml:85]: `quantus[torch]` should be enough

File: pyproject.toml [hunk 18]
```diff
@@ -85,7 +60,9 @@ torch = [
     "torchvision<=0.12.0; python_version == '3.7'",
     "torchvision>=0.15.1; sys_platform != 'linux' and python_version > '3.7'",
     "torchvision>=0.14.0, <0.15.1; sys_platform == 'linux' and python_version > '3.7' and python_version <= '3.10'",
-    "torchvision>=0.15.1; sys_platform == 'linux' and python_version >= '3.11'"
+    "torchvision>=0.15.1; sys_platform == 'linux' and python_version >= '3.11'",
+    "transformers<=4.30.2; python_version == '3.7'",
```
Comment (aaarrti) [inline pyproject.toml:64]: please remove `transformers` from `torch = [...]` section

**Review by aaarrti — CHANGES_REQUESTED** [review]
Please make sure Quantus is usable without `transformers` installation
